Creating library of human exons
================

## Prepare the environment

``` r
library(tidyverse)
```

    ## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
    ## ✔ dplyr     1.1.1     ✔ readr     2.1.4
    ## ✔ forcats   1.0.0     ✔ stringr   1.5.0
    ## ✔ ggplot2   3.4.2     ✔ tibble    3.2.1
    ## ✔ lubridate 1.9.2     ✔ tidyr     1.3.0
    ## ✔ purrr     1.0.1     
    ## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
    ## ✖ dplyr::filter() masks stats::filter()
    ## ✖ dplyr::lag()    masks stats::lag()
    ## ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors

We downloaded the human reference genome and the human gene annotation
from <https://www.gencodegenes.org/human/release_38.html>.

- **Genome sequence (GRCh38.p13)**
- **Comprehensive gene annotation**

Package used: **BEDOPS**. Before conversion, extract the file.

    gtf2bed < /Volumes/Temp1/Exons/gencode.v35.annotation.gtf > /Volumes/Temp1/Exons/gencode.v35.annotation.bed

## Exon library preparation

We select only the exons and add to the output file the name of the
gene. Then remove the “;” at the end of each line with **sed**. Then
select only exons from PCGs.

    awk '{ if ($8 == "exon") print $1 "\t" $2 "\t" $3 "\t" $4 "\t" $15 "\t" $23}' /Volumes/Temp1/human-genes/gencode.v38.annotation.bed > /Volumes/Temp1/human-genes/exons_raw.bed
    sed 's/;$//' /Volumes/Temp1/Exons/exons_raw.bed > /Volumes/Temp1/Exons/exons.bed
    awk '$5 == "protein_coding"' /Volumes/Temp1/Exons/exons.bed > /Volumes/Temp1/human-genes/exons/exons-pcg.bed

Then, using the python script **merge-overlapping-exons**, we merge
overlapping exons if they are part of the same gene, like in the case of
different isoforms. Exons overlapping but from different genes were not
merged with each other, like it would have occurred just by running
bedtools merge on the whole exon library.

The script filter the exon library by keeping only exons from one gene
at the time, and then run `bedtools merge` to merge the overlapping
exons. Then, it concatenates all the files in a single output file
(`merged-exons-pcg.bed`).

The final file has for columns, with each row representing one exon:

- `chromosome` -\> chromosome where the gene is located.
- `start` -\> starting position of the exon.
- `end` -\> ending position of the exon.
- `gene` -\> gene of which the exon is part of.
