---
title: "Creating library of human exons and genes"
output: rmarkdown::github_document
editor_options: 
  markdown: 
    wrap: sentence
---

## Prepare the environment

```{r env}
library(tidyverse)
library(ggpubr)
library(Biostrings)
```

Download the human reference genome and the human gene annotation from <https://www.gencodegenes.org/human/release_38.html>.

* **Genome sequence (GRCh38.p13)**
* **Comprehensive gene annotation**

Package used: **BEDOPS**. Before conversion, extract the file.
```
gtf2bed < /Volumes/Temp1/Exons/gencode.v35.annotation.gtf > /Volumes/Temp1/Exons/gencode.v35.annotation.bed
```

## Exon library preparation

### All exons

Select only the exons and add to the output file the name of the gene. Then remove the ";" at the end of each line with **sed**.
```
awk '{ if ($8 == "exon") print $1 "\t" $2 "\t" $3 "\t" $4 "\t" $23}' /Volumes/Temp1/Exons/gencode.v35.annotation.bed > /Volumes/Temp1/Exons/exons_raw.bed

sed 's/;$//' /Volumes/Temp1/Exons/exons_raw.bed > /Volumes/Temp1/Exons/exons.bed
```

With the created file with exons and chromosomal position, create a fasta file extracting the sequences from the reference genome file (extracted).
```
bedtools getfasta -fi /Volumes/Temp1/Exons/GRCh38.p13.genome.fa -bed /Volumes/Temp1/Exons/exons.bed -name -split -s -fo /Volumes/Temp1/Exons/exons.fa
```

```{r}
library <- readDNAStringSet("/Volumes/Temp1/human-genes/exons/exons.fa")
names <- names(library) %>% as_tibble()
df <- as.data.frame(library) %>% as_tibble() %>% mutate(length = nchar(as.character(x))) %>% select(length)
exons <- bind_cols(names, df) %>% dplyr::rename("sequence"=value) %>% arrange(desc(length))

ggplot(exons, aes(length))+
  geom_histogram(bins = 100)+
  xlim(0, 1000)+
  geom_vline(xintercept = 150, col = "red")+
  ggtitle("Exons count by length")+
  labs(subtitle = "All human exons")+
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5))

all <- exons %>% summarise(count = n()) %>% pull
lost <- exons %>% filter(length<150) %>% summarise(count = n()) %>% pull
all_bases <- exons %>% summarise(length = sum(length)) %>% pull
lost_bases <- exons %>% filter(length<150) %>% summarise(length = sum(length)) %>% pull

(percentage_exons_lost = (lost/all)*100)
(percentage_bases_lost = (lost_bases/all_bases)*100)
```


59% of the exons are shorter than the HGDP read length (150 bp), representing 21% of all the bases in the human exons.

### Only protein-coding-genes exons

```
awk '{ if ($8 == "exon") print $1 "\t" $2 "\t" $3 "\t" $4 "\t" $15 "\t" $23}' /Volumes/Temp1/human-genes/gencode.v38.annotation.bed > /Volumes/Temp1/human-genes/exons-pcg_raw.bed

sed 's/\"//g; s/;//g' /Volumes/Temp1/human-genes/exons/exons-pcg_raw.bed > /Volumes/Temp1/human-genes/exons/exons-pcg.bed

awk '$5 == "protein_coding"' /Volumes/Temp1/human-genes/exons/exons-all.bed > /Volumes/Temp1/human-genes/exons/exons-pcg.bed
```

```{r}
library_exons_pcg <- readDNAStringSet("/Volumes/Temp1/human-genes/exons/exons-pcg.fa")
names_exons_pcg <- names(library_exons_pcg) %>% as_tibble()
df_exons_pcg <- as.data.frame(library_exons_pcg) %>% as_tibble() %>% mutate(length = nchar(as.character(x))) %>% select(length)
exons_exons_pcg <- bind_cols(names_exons_pcg, df_exons_pcg) %>% dplyr::rename("sequence"=value) %>% arrange(desc(length))

ggplot(exons_exons_pcg, aes(length))+
  geom_histogram(bins = 100)+
  xlim(0, 1000)+
  geom_vline(xintercept = 150, col = "red")+
  ggtitle("Exons count by length")+
  labs(subtitle = "Only protein coding genes exons")+
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5))

all_exons_pcg <- exons_exons_pcg %>% summarise(count = n()) %>% pull
lost_exons_pcg <- exons_exons_pcg %>% filter(length<150) %>% summarise(count = n()) %>% pull
all_bases_exons_pcg <- exons_exons_pcg %>% summarise(length = sum(length)) %>% pull
lost_bases_exons_pcg <- exons_exons_pcg %>% filter(length<150) %>% summarise(length = sum(length)) %>% pull

(percentage_exons_lost_exons_pcg = (lost_exons_pcg/all_exons_pcg)*100)
(percentage_bases_lost_exons_pcg = (lost_bases_exons_pcg/all_bases_exons_pcg)*100)
```

Only considering protein coding genes exons, the percentage of lost sequences remains similar (61%) as well as the percentage of lost bases (23%).

## Gene library preparation

### All genes

```
awk '{ if ($8 == "gene") print $1 "\t" $2 "\t" $3 "\t" $4 "\t" $13}' /Volumes/Temp1/human-genes/gencode.v38.annotation.bed > /Volumes/Temp1/human-genes/genes_raw.bed

sed 's/\"//g; s/;//g' /Volumes/Temp1/human-genes/genes_raw.bed > /Volumes/Temp1/human-genes/genes/genes.bed
```

```{r}
library_genes <- readDNAStringSet("/Volumes/Temp1/human-genes/genes/genes.fa")
names_genes <- names(library_genes) %>% as_tibble()
df_genes <- as.data.frame(library_genes) %>% as_tibble() %>% mutate(length = nchar(as.character(x))) %>% select(length)
genes <- bind_cols(names_genes, df_genes) %>% dplyr::rename("sequence"=value) %>% arrange(desc(length))

ggplot(genes, aes(length))+
  geom_histogram(bins = 150)+
  xlim(0, 5000)+
  geom_vline(xintercept = 150, col = "red")+
  ggtitle("Gene count by length")+
  labs(subtitle = "All human genes")+
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5))

(all_genes <- genes %>% summarise(count = n()) %>% pull)
lost_genes <- genes %>% filter(length<150) %>% summarise(count = n()) %>% pull
all_bases_genes <- genes %>% summarise(length = sum(length)) %>% pull
lost_bases_genes <- genes %>% filter(length<150) %>% summarise(length = sum(length)) %>% pull

(percentage_genes_lost = (lost_genes/all_genes)*100)
(percentage_bases_lost_genes = (lost_bases_genes/all_bases_genes)*100)
```

There are 60.000 genes sequences in the library.
10% of the genes are shorter than the HGDP read length (150 bp), representing only the 0.03% of all the bases in the human genes.

### Only protein coding genes

```{r}
all_genes <- read_tsv("/Volumes/Temp1/human-genes/genes/genes.bed", col_names = c("chr", "start", "end", "name", "type"))

(per_type <- all_genes %>% group_by(type) %>% dplyr::summarise(count=n()))

format_numeric <- function(x) {
  formatC(x, format = "f", digits = 0)
}

protein_coding <- all_genes %>% filter(type == "protein_coding") %>% select(-type)

formatted_tbl <- protein_coding %>% 
  mutate_if(is.numeric, format_numeric)

#write.table(formatted_tbl, file = "/Volumes/Temp1/human-genes/genes/protein-coding-genes.bed", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

```{r}
library_pcg <- readDNAStringSet("/Volumes/Temp1/human-genes/genes/protein-coding-genes.fa")
names_pcg <- names(library_pcg) %>% as_tibble()
df_pcg <- as.data.frame(library_pcg) %>% as_tibble() %>% mutate(length = nchar(as.character(x))) %>% select(length)
pcg <- bind_cols(names_pcg, df_pcg) %>% dplyr::rename("sequence"=value) %>% arrange(desc(length))

ggplot(pcg, aes(length))+
  geom_histogram(bins = 150)+
  xlim(0, 5000)+
  geom_vline(xintercept = 150, col = "red")+
  ggtitle("Gene count by length")+
  labs(subtitle = "Only protein coding genes")+
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5))

(all_pcg <- pcg %>% summarise(count = n()) %>% pull)
lost_pcg <- pcg %>% filter(length<150) %>% summarise(count = n()) %>% pull
all_bases_pcg <- pcg %>% summarise(length = sum(length)) %>% pull
lost_bases_pcg <- pcg %>% filter(length<150) %>% summarise(length = sum(length)) %>% pull

(percentage_pcg_lost = (lost_pcg/all_pcg)*100)
(percentage_bases_lost_pcg = (lost_bases_pcg/all_bases_pcg)*100)
```

Considering only the **protein coding genes**, the fraction below 150 bp is really non significant (0.07%  for even less % of bases lost). In total, we have 19955 annotated protein coding genes.


## Overlapping sequences

```{r}
overlap <- function(bed, seq_type, cut) {

bed <- bed %>% filter(end-start>cut)
total_count <- nrow(bed) %>% as_tibble() %>% dplyr::rename("sequences" = value)

overlaps <- bed %>% mutate(prev_end = lag(end, default = 0), next_start = lead(start, default = Inf)) %>% filter(start < prev_end | end > next_start)
  
bases <- overlaps %>% mutate(overlap_begin = ifelse(prev_end-start>0, prev_end-start, 0), overlap_end = ifelse(end-next_start>0, end-next_start, 0), overlap = overlap_begin+overlap_end) %>% select(-overlap_begin, -overlap_end, -prev_end, -next_start)
  
overlap_count <- nrow(overlaps) %>% as_tibble() %>% dplyr::rename("overlapping_sequences" = value) %>% mutate(cutoff = cut)

tibble <- bind_cols(total_count, overlap_count) %>% mutate(percentage_overlapping = (overlapping_sequences/sequences)*100, type = seq_type) %>% relocate(cutoff, before = NULL) %>% relocate(type, before = NULL)

figure <- ggplot(bases, aes(overlap))+
  geom_histogram(bin = 500)+
  xlim(0, 1000)+
  ggtitle(seq_type)+
  labs(subtitle = paste0("Removed sequences shorter than ", cut))+
  xlab("Overlap (bp)")+
  theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5))

list(table = tibble, plot = figure)
}
```

```{r}
# All exons
exons <- read_tsv("/Volumes/Temp1/human-genes/exons/exons.bed", col_names = c("chr", "start", "end", "name", "type")) %>% select(-type) %>% distinct()
ov_ex <- overlap(exons, "all_exons", 0)
ov_ex_c <- overlap(exons, "all_exons", 150)

# PCG exons
exons_pcg <- read_tsv("/Volumes/Temp1/human-genes/exons/exons-pcg.bed", col_names = c("chr", "start", "end", "name", "type")) %>% select(-type, -X6) %>% distinct()
ov_ex_pcg <- overlap(exons_pcg, "pcg_exons", 0)
ov_ex_pcg_c <- overlap(exons_pcg, "pcg_exons", 150)

# All genes
genes <- read_tsv("/Volumes/Temp1/human-genes/genes/genes.bed", col_names = c("chr", "start", "end", "name", "type")) %>% distinct()
ov_g <- overlap(genes, "all_genes", 0)
ov_g_c <- overlap(genes, "all_genes", 150)

# PC genes
genes_pcg <- read_tsv("/Volumes/Temp1/human-genes/genes/protein-coding-genes.bed", col_names = c("chr", "start", "end", "name", "type")) %>% distinct()
ov_pcg <- overlap(genes_pcg, "pc_genes", 0)
ov_pcg_c <- overlap(genes_pcg, "pc_genes", 150)

# Final table
(tab <- bind_rows(ov_ex$table, ov_ex_c$table, ov_ex_pcg$table, ov_ex_pcg_c$table, ov_g$table, ov_g_c$table, ov_pcg$table, ov_pcg_c$table))

# Final plot (only with cutoff at 150)
(fig <- ggarrange(ov_ex_c$plot, ov_ex_pcg_c$plot, ov_g_c$plot, ov_pcg_c$plot))
```




