BLAST matrix preparation
================

## Prepare the environment

``` r
library(tidyverse)
```

    ## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
    ## ✔ dplyr     1.1.1     ✔ readr     2.1.4
    ## ✔ forcats   1.0.0     ✔ stringr   1.5.0
    ## ✔ ggplot2   3.4.2     ✔ tibble    3.2.1
    ## ✔ lubridate 1.9.2     ✔ tidyr     1.3.0
    ## ✔ purrr     1.0.1     
    ## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
    ## ✖ dplyr::filter() masks stats::filter()
    ## ✖ dplyr::lag()    masks stats::lag()
    ## ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors

From the annotation of exons with merged overlaps
**merged-exons-pcg.bed**, we extract the fasta sequences using
`bedtools getfasta` with the option `-name` to add the gene name to
every fasta header.

    bedtools getfasta -fi /Volumes/Temp1/human-genes/reference/ref-genome/GRCh38.p13.genome.fa -bed /Volumes/Temp1/human-genes/annotation/exons/merged-exons-pcg.bed -name+ > /Volumes/Temp1/human-genes/annotation/exons/merged-exons-pcg.fasta

We use the python script **create-transcripts** to merge the fasta
sequences of the non-overlapping exons from the same gene. In this way,
we create a library of “artificial transcripts” containing all the
non-overlapping exons.

    python /Volumes/Temp1/human-genes/scripts/library-prep/create-transcripts.py /Volumes/Temp1/human-genes/annotation/exons/merged-exons-pcg.fasta /Volumes/Temp1/human-genes/annotation/exons/transcripts.fasta

We use the resulting FASTA file of transcripts to calculate the
similarity of each transcript vs all the other transcripts in the
library using the script **blast.py**, which call the bash script
**blast-matrix.sh** to BLAST each FASTA entry to all the others. Then,
it calculates the similarity % based on BLAST hits on all the other
transcripts.

    python /Volumes/Temp1/human-genes/blast-matrix/blast.py /Volumes/Temp1/human-genes/annotation/exons/transcripts.fasta /Volumes/Temp1/human-genes/annotation/exons/transcripts.fasta /Volumes/Temp1/human-genes/blast-matrix/blast-transcripts/

The final output is this matrix:

``` r
(similarity_matrix <- read_tsv("/Volumes/Temp1/human-genes/blast-matrix/similarity.blast", col_names = c("gene", "similar_to", "percentage_identity", "equal_bases", "gene_length")))
```

    ## Rows: 13689 Columns: 5
    ## ── Column specification ────────────────────────────────────────────────────────
    ## Delimiter: "\t"
    ## chr (2): gene, similar_to
    ## dbl (3): percentage_identity, equal_bases, gene_length
    ## 
    ## ℹ Use `spec()` to retrieve the full column specification for this data.
    ## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

    ## # A tibble: 13,689 × 5
    ##    gene               similar_to     percentage_identity equal_bases gene_length
    ##    <chr>              <chr>                        <dbl>       <dbl>       <dbl>
    ##  1 ENSG00000186092.7  ENSG000001766…                96.4        2525        2618
    ##  2 ENSG00000186092.7  ENSG000001776…                40.7        1066        2618
    ##  3 ENSG00000284733.2  ENSG000002301…               100           939         939
    ##  4 ENSG00000284733.2  ENSG000002846…               100           939         939
    ##  5 ENSG00000284733.2  ENSG000001762…                99.7         936         939
    ##  6 ENSG00000284662.1  ENSG000002301…               100           995         995
    ##  7 ENSG00000284662.1  ENSG000002847…                94.4         939         995
    ##  8 ENSG00000284662.1  ENSG000001762…                94.1         936         995
    ##  9 ENSG00000187583.11 ENSG000001876…                19.2         662        3451
    ## 10 ENSG00000187642.9  ENSG000001875…                19.3         662        3424
    ## # ℹ 13,679 more rows

And this is the raw blast matrix, only keeping bitscore \> 1000. From
this, the script calculates the final matrix. This raw matrix might be
useful in the future, cause it contains lot of relevant infos.

BLAST output format: qseqid sseqid pident qlen length mismatch gapopen
qstart qend sstart send evalue bitscore

``` r
(raw_blast_matrix <- read_tsv("/Volumes/Temp1/human-genes/blast-matrix/merged.blast", col_names = c("gene", "similar_to", "percentage_identity", "match_length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "x", "bitscore")) %>% select(-x))
```

    ## Rows: 30355 Columns: 13
    ## ── Column specification ────────────────────────────────────────────────────────
    ## Delimiter: "\t"
    ## chr  (2): gene, similar_to
    ## dbl (11): percentage_identity, match_length, mismatch, gapopen, qstart, qend...
    ## 
    ## ℹ Use `spec()` to retrieve the full column specification for this data.
    ## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

    ## # A tibble: 30,355 × 12
    ##    gene      similar_to percentage_identity match_length mismatch gapopen qstart
    ##    <chr>     <chr>                    <dbl>        <dbl>    <dbl>   <dbl>  <dbl>
    ##  1 ENSG0000… ENSG00000…                99.6         2618     2535      10      0
    ##  2 ENSG0000… ENSG00000…                99.4         2618     1072       6      0
    ##  3 ENSG0000… ENSG00000…               100            939      939       0      0
    ##  4 ENSG0000… ENSG00000…               100            939      939       0      0
    ##  5 ENSG0000… ENSG00000…                99.7          939      939       3      0
    ##  6 ENSG0000… ENSG00000…               100            995      995       0      0
    ##  7 ENSG0000… ENSG00000…               100            995      939       0      0
    ##  8 ENSG0000… ENSG00000…                99.7          995      939       3      0
    ##  9 ENSG0000… ENSG00000…               100           3451      662       0      0
    ## 10 ENSG0000… ENSG00000…               100           3424      662       0      0
    ## # ℹ 30,345 more rows
    ## # ℹ 5 more variables: qend <dbl>, sstart <dbl>, send <dbl>, evalue <dbl>,
    ## #   bitscore <dbl>

To obtain the paralogs, we filter for gene with percentage of identity
\>80%, obtaining this final file, which is used in our pipeline to sum
up paralogs coverages:

``` r
(similarity80_matrix <- read_tsv("/Volumes/Temp1/human-genes/blast-matrix/similarity80.blast"))
```

    ## Rows: 2379 Columns: 5
    ## ── Column specification ────────────────────────────────────────────────────────
    ## Delimiter: "\t"
    ## chr (2): gene, similar_to
    ## dbl (3): pid, id_bases, gene_length
    ## 
    ## ℹ Use `spec()` to retrieve the full column specification for this data.
    ## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.

    ## # A tibble: 2,379 × 5
    ##    gene              similar_to          pid id_bases gene_length
    ##    <chr>             <chr>             <dbl>    <dbl>       <dbl>
    ##  1 ENSG00000186092.7 ENSG00000176695.8  96.4     2525        2618
    ##  2 ENSG00000284733.2 ENSG00000230178.1 100        939         939
    ##  3 ENSG00000284733.2 ENSG00000284662.1 100        939         939
    ##  4 ENSG00000284733.2 ENSG00000176269.4  99.7      936         939
    ##  5 ENSG00000284662.1 ENSG00000230178.1 100        995         995
    ##  6 ENSG00000284662.1 ENSG00000284733.2  94.4      939         995
    ##  7 ENSG00000284662.1 ENSG00000176269.4  94.1      936         995
    ##  8 ENSG00000116721.9 ENSG00000204481.8  97.5     2377        2439
    ##  9 ENSG00000116721.9 ENSG00000279169.3  97.4     2376        2439
    ## 10 ENSG00000239810.3 ENSG00000204501.7  97.9     1806        1845
    ## # ℹ 2,369 more rows
