---
title: "Pipeline workflow"
output: rmarkdown::github_document
editor_options: 
  markdown: 
    wrap: sentence
---

## Prepare the environment

```{r env}
library(tidyverse)
```


1. Call **mapping.sh**, which is taking as input a folder containing files in `fastq.gz` format and the `reference FASTA genome`. This is usually the format in which the files are downloaded from ENA. If the downloaded files are from paired-end sequencing, we first merge them together to have a single fastq file per sample.
The script is unzipping the file, mapping to a reference genome to obtain the `sam` file, converting it into `bam`, sorting into `sort.bam` and indexing it.
```
bash /Volumes/Temp1/human-genes/scripts/pipeline/mapping.sh /Volumes/Temp1/human-genes/playground/fastq /Volumes/Temp1/human-genes/reference/ref-genome/GRCh38.p13.genome.fa
```

2. Call **scg.coverage.sh**, which takes as input the `sorted.bam` file, the `scg.bed` file containing the genomic coordinates of selected single-copy-genes and output a `scg.tsv` file with the mean depth of each single copy gene.
```
bash /Volumes/Temp1/human-genes/scripts/pipeline/scg-coverage.sh /Volumes/Temp1/human-genes/playground/ERR4059675.sorted.bam /Volumes/Temp1/human-genes/annotation/genes/scg.bed /Volumes/Temp1/human-genes/playground/ERR4059675.scg.tsv
```

3. Call **pcg.coverage.sh**, which takes again the `sorted.bam` file, the `merged-exons-pcg.bed` file containing the genomic coordinates of protein coding gene exons, and output a `pcg.tsv` file with the mean depth of each exon and their corresponding gene name.
```
bash /Volumes/Temp1/human-genes/scripts/pipeline/pcg-coverage.sh /Volumes/Temp1/human-genes/playground/ERR4059675.sorted.bam /Volumes/Temp1/human-genes/annotation/exons/merged-exons-pcg.bed /Volumes/Temp1/human-genes/playground/ERR4059675.pcg.tsv
```

4. Call **pcg-merged-coverage.py**, which takes the `pcg.tsv` file and calculates the mean coverage of each gene, summing up the exons coverages weighted by their length and dividing by the gene length. Returns the `pcg-merged.csv` file, with coverage estimates for each gene and the length of the coding region (which IS NOT matching the genomic position of the gene in column 1, which includes the introns).
```
python /Volumes/Temp1/human-genes/scripts/pipeline/pcg-merged-coverage.py /Volumes/Temp1/human-genes/playground/ERR4059675.pcg.tsv /Volumes/Temp1/human-genes/playground/ERR4059675.pcg-merged.tsv
```

5. Call the script **normalize.py**, which takes the `scg.tsv` and `pcg.tsv`, calculates the *normalizer*, which is the mean coverage of the SCGs, and normalize the raw coverage of each PCG to get the normalized copynumber estimates in the file `normalized-coverage.tsv`.
```
python /Volumes/Temp1/human-genes/scripts/pipeline/normalize.py /Volumes/Temp1/human-genes/playground/ERR4059675.scg.tsv /Volumes/Temp1/human-genes/playground/ERR4059675.pcg-merged.tsv /Volumes/Temp1/human-genes/playground/ERR4059675-normalized-coverage.tsv
```

5. Call the script **gc-normalization.R**, which takes the `scg.tsv` and `normalized-coverage.tsv` files, the GC estimates for SCGs (`scg-gc.tsv`) and PCGs (`pcg-gc.tsv`) as inputs. It calculates the fitting coefficient of the parabola fitted to the GC-coverage correlation on SCGs, which has equation:

 \(coverage = a + b*GC + c*(GC^2)\)

With `GC`= percentage of GC in the given gene.
Then, it normalizes the coverage of each PCG based on its GC % using the equation:

 \(GC-normalized copynumber = copynumber + {[1-(a + b*GC + c*GC^2)]*copynumber}\)
 
The equation is:

1. Calculating which is the **expected coverage for 1 copy** of the gene based on its GC-content using the parabola equation.
2. Calculate the **difference between 1 and the expected coverage** (1 is the unbiased result, the expected coverage is the GC-biased result)
3. Multiplying this difference for the number of copies estimated from the SCG normalization for that PCG (the idea is that every estimated copy should be normalized by the GC-bias), to obtain the **"adjustment value"**.
4. Summing the estimated copynumber from the SCG normalization with the adjustment value to get the **GC-biased corrected copynumber**.
```
Rscript /Volumes/Temp1/human-genes/scripts/pipeline/gc-normalization.R --gc_scg=/Volumes/Temp1/human-genes/gc-normalization/scg-gc.tsv --cov_scg=/Volumes/Temp1/human-genes/playground/ERR4059675.scg.tsv --pcg_cn=/Volumes/Temp1/human-genes/playground/ERR4059675-normalized-coverage.tsv --gc_pcg=/Volumes/Temp1/human-genes/gc-normalization/pcg-gc.tsv --output=/Volumes/Temp1/human-genes/playground/
```


